Port VimL's method call syntax - vim-patch:8.1.{1638,1800,1803,1807,1809,1816,1820,1821,1828,1834,1835,1861,1863,1878,1879,1888,1909,1911,1912}

seandewar opened the pull request:
This is a WIP port of VimL's method call syntax: `base->method(...)`

Ref: #12525, #14611

I'm tempted to do this in multiple PRs to avoid a patch storm like #15211 to make reviewing this easier.
I was aiming for a relatively bug-free initial implementation in that PR which required _many_ patches.

However, method syntax seems to be quite bug-free from the get-go, with most patches only adding method call support to existing functions.

~~Note that I probably won't aim to support using `v:lua` as a method here.~~
Changed my mind 😄. `v:lua` may be used as a method like:
`base->v:lua.table.func(...)`

seandewar committed:
vim-patch:8.1.1800: function call functions have too many arguments

Problem:    Function call functions have too many arguments.
Solution:   Pass values in a funcexe_T struct.
https://github.com/vim/vim/commit/c6538bcc1cdd1fb83732f22fdc69bd9bb66f968a

Use FUNCEXE_INIT to initialize funcexe_T instances.

call_callback() and other Vim listener related stuff is N/A.

seandewar committed:
fix(eval): cherry-pick emsg changes from v8.1.0736

v8.1.0736 made some changes for making some emsgs more specific.
Includes the change for Test_lambda_fails() in test_lambda.vim.
Adjust relevant functionaltests to expect the new emsgs.

This patch has been fully ported in my Blob port PR, but it hasn't been
merged yet, so just use what we need from it for now.

Required for v8.1.1821.

seandewar committed:
vim-patch:8.1.1803: all builtin functions are global

Problem:    All builtin functions are global.
Solution:   Add the method call operator ->.  Implemented for a limited number
            of functions.
https://github.com/vim/vim/commit/ac92e25a33c37ec5becbfffeccda136c73b761ac

- Note that to *exactly* port hunk @@ -7376,18 +7444,19 from
  handle_subscript(), we need the :scriptversion patches (I have an open
  PR for those, but this patch works fine without them anyway).
- Port call_internal_func() from v7.4.2058.
- Adjust some error messages in tests, as they rely on the Blob patches.
- Add a modeline to test_method.vim.

Ignore the global_functions and base_method tables and prefer the
current GPerf implementation. Instead, add an extra base_arg field to
VimLFuncDef that holds the number of the argument to use as the base
(1-indexed, so that 0 may be used to refer to functions that cannot be
used as methods).

This also means we support using any argument as a base from the get-go,
rather than just the first (Vim includes this ability in future patches,
however).

To mark a function as usable as a method, use the "base" key as
described in eval.lua.

seandewar committed:
vim-patch:8.1.1807: more functions can be used as a method

Problem:    More functions can be used as a method.
Solution:   Add append(), appendbufline(), assert_equal(), etc.
            Also add the :eval command.
https://github.com/vim/vim/commit/25e42231d3ee27feec2568fa4be2aa2bfba82ae5

:eval is already ported.

seandewar commented:
@github-actions bad bot 👎

dundargoc commented:
> @github-actions bad bot 👎 

What did it do?

seandewar commented:
@dundargoc Nothing. I just didn't like its choice of labels is all 😄

bfredl commented:
Maybe it should not add topic labels if someone already did that :)

seandewar committed:
vim-patch:8.1.1809: more functions can be used as a method

Problem:    More functions can be used as a method.
Solution:   Add has_key(), split(), str2list(), etc.
https://github.com/vim/vim/commit/a74e4946de074d2916e3d6004f7fa1810d12dda9

seandewar committed:
vim-patch:8.1.1816: cannot use a user defined function as a method

Problem:    Cannot use a user defined function as a method.
Solution:   Pass the base as the first argument to the user defined function
            after "->". (partly by FUJIWARA Takuya)
https://github.com/vim/vim/commit/fcfe1a9b8950b8b211ab3b24d84b17c6847ea43f

seandewar committed:
vim-patch:8.1.1820: using expr->FuncRef() does not work

Problem:    Using expr->FuncRef() does not work.
Solution:   Make FuncRef work as a method.
https://github.com/vim/vim/commit/761fdf01c6e307c448cec2684f8b315ba6d1f454

clason commented:
Yeah, it's (also) a bit eager with the `documentation` label (and often calls it a day even when there's actual code in the PR this documentation is about...)

bfredl commented on src/nvim/eval.c:
```diff
@@ -8470,9 +8546,15 @@ handle_subscript(
-                        &len, evaluate, pt, selfdict);
+
+      funcexe_T funcexe;
+      memset(&funcexe, 0, sizeof(funcexe));
```
Instead of memset, i would suggest defining `FUNCEXE_INIT`  in analogy with existing types.

dundargoc commented:
> Maybe it should not add topic labels if someone already did that :)

Probably not worth it. The labeler is only triggered the first time a PR is opened. 

It takes around 10 seconds for the labels to be applied when I benchmarked it. Not only that, the author must be quick enough to label not only before the labeler is finished, but at the beginning of the labeler script where it's possible to check if there are any labels, so more like 5 seconds. I think it would only be relevant for label speedrunners.

The reason this particular PR didn't manage to label in time is because the GitHub servers tanked today for some reason. [Take a look at the recent workflow times](https://github.com/neovim/neovim/actions/workflows/labeler.yml). It took 25 min to label this PR. It's an anomaly that I don't think we should take into consideration for our workflow.

I propose what we do instead is refine and pinpoint when we want to apply labels. As I see it bot labeled correctly: the documentation files were modified so it added the `documentation` label.  The files under src/nvim/lua were also modified so the `lua` label was also applied. What makes these labels unsuited in this case as opposed to others when these files are modified? Is it because it's a vim-patch, and that vim-patch shouldn't activate the normal labels?

bfredl commented:
@dundargoc every feature PR should add its own documentation. If "documentation" tag is to be added to every PR which is not a bugfix, the label is worthless.

dundargoc commented:
> @dundargoc every feature PR should add its own documentation. If "documentation" tag is to be added to every PR which is not a bugfix, the label is worthless.

Agreed! That's why I think we should redefine when we want to use the documentation label. [I've created a PR for this](https://github.com/neovim/neovim/pull/15298), I propose we move the discussion there so @seandewar can work on his PR in peace :) .

seandewar committed:
vim-patch:8.1.1821: no test for wrong number of method arguments

Problem:    No test for wrong number of method arguments.
Solution:   Add a test.
https://github.com/vim/vim/commit/f97d46f816e84edb6899a903a1c334a50a6d31bb

seandewar committed:
vim-patch:8.1.1828: not strict enough checking syntax of method invocation

Problem:    Not strict enough checking syntax of method invocation.
Solution:   Check there is no white space inside ->method(.
https://github.com/vim/vim/commit/5184132ec015f5889a3195d911e609d214f06bed

seandewar committed:
vim-patch:8.1.1834: cannot use a lambda as a method

Problem:    Cannot use a lambda as a method.
Solution:   Implement ->{lambda}(). (closes vim/vim#4768)
https://github.com/vim/vim/commit/22a0c0c4ecd23b6c43f79ba9b92899ca0b426e29

Add an additional lua_funcname argument to call_func_rettv() to maintain
support for v:lua.

A memory leak was introduced with this patch that was fixed in
v8.1.2107.

seandewar committed:
fix(eval_lambda): cherry-pick leak fix from v8.1.2107

That patch also includes a test using test_refcount() for lambdas, but
such test functions are N/A for Nvim.

seandewar committed:
vim-patch:8.1.1835: cannot use printf() as a method

Problem:    Cannot use printf() as a method.
Solution:   Pass the base as the second argument to printf().
https://github.com/vim/vim/commit/fd8ca21b3ff207e44891aef922935d4adcd140cf

seandewar committed:
vim-patch:8.1.1861: only some assert functions can be used as a method

Problem:    Only some assert functions can be used as a method.
Solution:   Allow using most assert functions as a method.
https://github.com/vim/vim/commit/24278d2407dfbc8d93eb36593cdd006ff5d86f94

Port tests to assert_spec.lua.

seandewar committed:
vim-patch:8.1.1863: confusing error when using a builtin function as method

Problem:    Confusing error when using a builtin function as method while it
            does not support that.
Solution:   Add a specific error message.
https://github.com/vim/vim/commit/9174639a82799011cfa0013cbc4c4709b3833bf0

seandewar committed:
vim-patch:8.1.1878: negative float before method not parsed correctly

Problem:    Negative float before method not parsed correctly.
Solution:   Apply "!" and "-" in front of expression before using ->.
https://github.com/vim/vim/commit/9cfe8f6e68de4bfb5942d84f4465de914a747b3f

seandewar committed:
vim-patch:8.1.1879: more functions can be used as methods

Problem:    More functions can be used as methods.
Solution:   Make float functions usable as a method.
https://github.com/vim/vim/commit/93cf85f9ef02931de3f8c8e536a137da0b48b7dc

Fix atan2() doc typo (patch referred to it as atan()).
Adjust Test_fmod() method test to expect "str2float('nan')".

seandewar committed:
vim-patch:8.1.1888: more functions can be used as methods

Problem:    More functions can be used as methods.
Solution:   Make various functions usable as a method.
https://github.com/vim/vim/commit/073e4b92e613d22ce7b16e0fbf5c0e40cb5f9b2c

test_popup.vim already has the changes from this patch (they're N/A
anyway).

seandewar committed:
vim-patch:8.1.1638: running tests leaves some files behind

Problem:    Running tests leaves some files behind.
Solution:   Delete the files. (Ozaki Kiichi, closes vim/vim#4617)
https://github.com/vim/vim/commit/3940ec6d41a07f9abbfba7d4db6b49d3d8b43a9a

Required for v8.1.1888.

N/A patches for version.c:

vim-patch:8.1.1912: more functions can be used as methods

Problem:    More functions can be used as methods.
Solution:   Make channel and job functions usable as a method.
vim/vim@570497a

seandewar committed:
vim-patch:8.1.1909: more functions can be used as methods

Problem:    More functions can be used as methods.
Solution:   Make a few more functions usable as a method.
https://github.com/vim/vim/commit/e49fbff384e45dd17fed72321c26937edf6de16b

seandewar committed:
vim-patch:8.1.1911: more functions can be used as methods

Problem:    More functions can be used as methods.
Solution:   Make a few more functions usable as a method.
https://github.com/vim/vim/commit/64b4d73524b9a2304d89b87529cd8d3cef14b856

Note that the old-style version of Test_byteidx() was already translated
to a Lua test in 069_multibyte_formatting_spec.lua. Keep both versions,
using Test_byteidx() to mainly test the method call syntax for byteidx()
and byteidxcomp().

seandewar committed:
test(oldtest): cherry-pick needed test functions from v8.1.1336

Cherry-pick Test_call() and Test_cindent_func().

Required for v8.1.1911 and v8.1.1915.

seandewar commented on src/nvim/testdir/test_functions.vim:
```diff
@@ -875,6 +875,34 @@ func Test_byte2line_line2byte()
   bw!
 endfunc
 
+func Test_byteidx()
```
Ideally, how should this test be included?

It's a new-style version of a test in test69 that was already lua'd here: https://github.com/neovim/neovim/blob/35041432bff6812126e1cc04d30d35a5c9117537/test/functional/legacy/069_multibyte_formatting_spec.lua#L309

Difference though is that this test uses method call syntax, which is why I'd like to include it.

As the patch removed the old version of the test, should I remove the lua'd version?
Should I keep both versions of the test? Should I translate this to a lua test? :shrug:

bfredl replied:
I'll keep both in this case. the lua test tests the actual functionality, and this one for the `->` shenanigans

seandewar committed:
feat(eval): partially port v8.1.1915

Cannot be fully ported as chdir() hasn't been ported yet.

seandewar committed:
feat(v:lua): support calling v:lua as a method

seandewar committed:
fix(v:lua): fix emsg when calling v:lua directly

v:lua expressions are represented using vvlua_partial. As v:lua isn't
intended to be called directly, it's given an empty pt_name.

Because of this, calling v:lua directly like "v:lua()" will cause "E117:
Unknown function: ", with an empty name.

Instead, have call_func() show the name "v:lua" in the emsg.

seandewar commented:
I think this is ready for review.

This PR doesn't include _all_ method-related patches.
The other patches (barring Vim9 stuff) add method support to existing built-in functions.
Those can be ported pretty easily in future PRs by editing `eval.lua` (I want to avoid making this PR too big).

Also, no Nvim-only functions have been made methods in this PR.

Btw, in the end I decided to add support for using `v:lua` as a method in this PR, e.g:

```vim
[[1], [2, 3], [4]]->v:lua.vim.tbl_flatten()->map({_, v -> v * 5})
" [5, 10, 15, 20]
```

bfredl commented:
I haven't looked at everything, but the changes around `v:lua` LGTM.

justinmk reviewed (approved):
This is a thoughtful and careful port, thank you.

Just to be clear, we explicitly will not support Vimscript 2 (Vim9 vimscript). PRs adding such support will be rejected, we want to focus on Lua and editor features.
